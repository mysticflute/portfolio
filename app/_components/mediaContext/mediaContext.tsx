'use client';

import { createContext, useContext, useReducer, Dispatch } from 'react';

/**
 * Describes the media currently playing on the page.
 */
export type State = {
  /** The id of the current audio or video currently playing on the page. */
  currentTrackId: string | null;
};

/**
 * Describes a change to the media currently playing on the page.
 */
export type Action = { type: 'playing'; id: string };

const StateContext = createContext<State | undefined>(undefined);
const DispatchContext = createContext<Dispatch<Action> | undefined>(undefined);

// handles dispatch calls
function reducer(state: State, action: Action): State {
  switch (action.type) {
    case 'playing':
      return {
        ...state,
        currentTrackId: action.id,
      };
    default: {
      throw new Error(`MediaContext: unhandled action type "${action.type}"`);
    }
  }
}

/**
 * A React context provider for reading and providing info
 * about the currently playing media on the page.
 */
export function MediaContextProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [currentMedia, dispatch] = useReducer(reducer, {
    currentTrackId: null,
  });

  return (
    <StateContext.Provider value={currentMedia}>
      <DispatchContext.Provider value={dispatch}>
        {children}
      </DispatchContext.Provider>
    </StateContext.Provider>
  );
}

/**
 * Gets the context for the currently playing media on the page.
 */
export function useMediaContext(): State {
  const context = useContext(StateContext);
  if (context === undefined) {
    throw new Error('useMediaContext must be used within MediaProvider');
  }
  return context;
}

/**
 * Gets the context to dispatch an update about the
 * currently playing media on the page.
 */
export function useMediaDispatch(): Dispatch<Action> {
  const dispatch = useContext(DispatchContext);
  if (dispatch === undefined) {
    throw new Error('useMediaDispatch must be used within MediaProvider');
  }
  return dispatch;
}
