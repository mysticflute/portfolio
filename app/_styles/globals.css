/* The file contains root-level custom properties and element resets only. */

:root {
  /* colors */
  --color-gray-100: #0b0b0b;
  --color-gray-200: #393939;
  --color-gray-300: #5e5e5f;
  --color-gray-400: #878787;
  --color-neutral-100: #fff;
  --color-neutral-200: #eff0f6;
  --color-neutral-300: #eceae8;
  --color-red-100: #e85d75;
  /* --color-red-100: #e574bc; */
  --color-yellow-100: #f6b468;
  --color-blue-100: #53abf8;
  --color-purple-100: #4331f0;
  --color-purple-200: #5358f8;
  --color-primary: var(--color-purple-100);
  --color-accent-100: var(--color-purple-200);
  --color-accent-200: var(--color-blue-100);
  --color-accent-300: var(--color-yellow-100);
  --color-accent-400: var(--color-red-100);
  --color-background-100: var(--color-gray-100);
  --color-background-200: var(--color-neutral-100);
  --color-background-300: var(--color-gray-100);
  --color-background-highlight: var(--color-blue-100);
  --color-text: var(--color-gray-200);
  --color-text-accent: var(--color-gray-100);
  --color-text-alt: var(--color-gray-100);
  --color-text-link: var(--color-gray-100);
  --color-text-highlight: var(--color-blue-100);
  --color-text-inverse: var(--color-neutral-200);

  /* spacing */
  --layout-max-width: 1300px;
  --layout-padding-x: 24px;
  --layout-padding-y: 0;
  --layout-padding: var(--layout-padding-y) var(--layout-padding-x);
  --layout-margin-x: auto;
  --layout-margin-y: 0;
  --layout-margin: var(--layout-margin-y) var(--layout-margin-x);
}

*,
::before,
::after {
  box-sizing: border-box;
}

html {
  scroll-behavior: smooth;
  scroll-padding: 2rem;
}

body {
  color: var(--color-text);
  background: var(--color-primary);
  font-family: var(--font-default), system-ui, sans-serif;
  font-weight: 500;
  font-size: 1.125rem;
  line-height: 1.667em;
}

@media screen and (max-width: 767px) {
  body {
    font-size: 1rem;
  }
}

h1,
h2,
h3,
h4,
h5,
h6 {
  margin: 0;
  padding: 0;
  font-weight: inherit;
  font-size: 1rem;
}

ol,
ul {
  margin: 0;
  padding: 0;
  list-style: none;
}

p {
  margin: 0;
  padding: 0;
}

audio,
canvas,
iframe,
img,
svg,
video {
  vertical-align: middle;
}

img {
  max-width: 100%;
  height: auto;
}

a {
  color: inherit;
  text-decoration: none;
}

@media screen and (max-width: 991px) {
  :root {
    --layout-max-width: calc(600px + var(--layout-padding-x) * 2);
  }
}

@media screen and (max-width: 767px) {
  :root {
    --layout-max-width: calc(500px + var(--layout-padding-x) * 2);
  }
}

@media screen and (max-width: 479px) {
  :root {
    --layout-padding-x: 20px;
  }
}
