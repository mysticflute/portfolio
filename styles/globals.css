:root {
  --color-dark: #0b0b0b;
  --color-blue: #1c92ff;
  --color-red: #ff4a60;
  --color-yellow: #ffc431;
  --color-text-default: #393939;
  --color-text-dark: var(--color-dark);
  --color-text-header-link: var(--color-dark);
  --color-text-header-link-highlight: var(--color-blue);
  --color-background-default: #faf9f6;
  --color-background-dark: var(--color-dark);
  --color-background-highlight: var(--color-blue);
}

*,
::before,
::after {
  box-sizing: border-box;
}

body {
  color: var(--color-text-default);
  background: var(--color-background-default);
  font-family: var(--font-default), system-ui, sans-serif;
  font-weight: 500;
  font-size: 1.125rem;
  line-height: 1.667em;
}

@media screen and (max-width: 767px) {
  body {
    font-size: 1rem;
  }
}

h1,
h2,
h3,
h4,
h5,
h6 {
  margin: 0;
  padding: 0;
  font-weight: inherit;
  font-size: 1rem;
}

ol,
ul {
  margin: 0;
  padding: 0;
  list-style: none;
}

p {
  margin: 0;
  padding: 0;
}

audio,
canvas,
iframe,
img,
svg,
video {
  vertical-align: middle;
}

img {
  max-width: 100%;
  height: auto;
}

a {
  color: inherit;
  text-decoration: none;
}

/* @media (prefers-color-scheme: dark) {
  html {
    color-scheme: dark;
  }
} */

/* @media (prefers-color-scheme: dark) {
  :root {
  }
} */
