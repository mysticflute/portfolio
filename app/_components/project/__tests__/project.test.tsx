import { describe, expect, it } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import { MediaContextProvider } from '@/components/mediaContext/mediaContext';
import Project from '../project';

describe('project', () => {
  it('matches snapshot', () => {
    render(
      <MediaContextProvider>
        <Project
          project={{
            name: 'Test Project',
            slug: 'test-project',
            sort: 1,
            role: 'Composer',
            description: 'A test project.',
            icon: 'pulsar/leaf.svg',
            link: 'https://www.test-project.com',
            color: '#fffeee',
            tracks: [
              { id: '123456', name: 'Test Track', mp3: 'https://url.mp3' },
            ],
          }}
        ></Project>
      </MediaContextProvider>,
    );

    expect(screen.getByRole('article')).toMatchSnapshot();
  });

  it('does not render an icon if not specified', () => {
    render(
      <Project
        project={{
          name: 'Test Project',
          slug: 'test-project',
          sort: 1,
          role: 'Composer',
          description: 'A test project.',
          link: 'https://www.test-project.com',
          color: '#fffeee',
        }}
      ></Project>,
    );

    expect(screen.queryByRole('presentation')).not.toBeInTheDocument();
  });

  it('does not render a project link if not specified', () => {
    render(
      <Project
        project={{
          name: 'Test Project',
          slug: 'test-project',
          sort: 1,
          role: 'Composer',
          description: 'A test project.',
        }}
      ></Project>,
    );

    expect(
      screen.queryByRole('link', { name: 'View project website' }),
    ).not.toBeInTheDocument();
  });

  it('does not render tracks if not specified', () => {
    render(
      <MediaContextProvider>
        <Project
          project={{
            name: 'Test Project',
            slug: 'test-project',
            sort: 1,
            role: 'Composer',
            description: 'A test project.',
            icon: 'pulsar/leaf.svg',
            link: 'https://www.test-project.com',
            color: '#fffeee',
            tracks: undefined,
          }}
        ></Project>
      </MediaContextProvider>,
    );

    expect(screen.queryByTestId('tracks')).not.toBeInTheDocument();
  });
});
