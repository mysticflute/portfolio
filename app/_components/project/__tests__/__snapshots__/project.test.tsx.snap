// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project matches snapshot 1`] = `
<article
  class="box hover container"
  style="--color-project: #fffeee;"
>
  <div
    class="info"
  >
    <img
      alt=""
      data-nimg="1"
      decoding="async"
      height="46"
      loading="lazy"
      src="/images/icons/pulsar/leaf.svg"
      style="color: transparent;"
      width="46"
    />
    <div
      class="badge"
    >
      Composer
    </div>
  </div>
  <h3
    class="textHeadingSmall"
  >
    Test Project
  </h3>
  <p>
    A test project.
  </p>
  <div
    class="tracks"
    data-testid="tracks"
  >
    <div
      class="track"
      data-testid="audio-track"
    >
      <button
        class="title"
        title="Test Track"
      >
        Test Track
      </button>
      <div
        class="audio"
      >
        <div
          aria-label="Audio player"
          class="rhap_container rhap_loop--off rhap_play-status--paused "
          role="group"
          tabindex="0"
        >
          <audio
            preload="none"
            src="https://url.mp3"
          />
          <div
            class="rhap_main rhap_horizontal"
          >
            <div
              class="rhap_progress-section"
            >
              <div
                class="rhap_time rhap_current-time"
                id="rhap_current-time"
              >
                --:--
              </div>
              <div
                aria-label="Audio progress control"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="0"
                class="rhap_progress-container"
                role="progressbar"
                tabindex="0"
              >
                <div
                  class="rhap_progress-bar rhap_progress-bar-show-download"
                >
                  <div
                    class="rhap_progress-indicator"
                    style="left: 0%;"
                  />
                  <div
                    class="rhap_progress-filled"
                    style="width: 0%;"
                  />
                </div>
              </div>
              <div
                class="rhap_time rhap_total-time"
              >
                --:--
              </div>
            </div>
            <div
              class="rhap_controls-section"
            >
              <div
                class="rhap_additional-controls"
              />
              <div
                class="rhap_main-controls"
              >
                <button
                  aria-label="Play"
                  class="rhap_button-clear rhap_main-controls-button rhap_play-pause-button"
                  type="button"
                >
                  <span />
                </button>
              </div>
              <div
                class="rhap_volume-controls"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="viewMore"
  >
    <a
      href="https://www.test-project.com"
    >
      View project website
      <span
        aria-hidden="true"
        class="lineRounded arrowDiagonal hasTextBefore arrow"
        role="img"
      />
    </a>
  </div>
</article>
`;
